### DISCUSSION TITLE
User defined semantic for shader resources

### DISCUSSION THREAD

**Discussion by Rodousse** [shader-slang/slang#3758] [Q&A] [answered]
Hello everyone,

First of all thank you for this great shader language initiative. I discovered the project with your vulkanised conference and the first few hours spent using the language were a real pleasure, I never got excited this much when programming shader!  :smile:

So I have a simple question concerning the [shader binding semantics](https://github.com/shader-slang/slang/blob/master/docs/user-guide/02-conventional-features.md#binding-semantics). It seems the binding semantics are usable on shader stage inputs and outputs, which might be system defined and even user defined.  
Now I tried this feature with shader resources or constant buffers as following:
```
RWTexture2D<float4> out_image : SV_DestinationTexture;
ConstantBuffer<float> time: SV_Time;
```
If I add these resources in a larger shader code, it will compile, link and give a valid output. Although, when using your reflection API the semantic name or index for these variables will be null.

It is pretty easy at the scale of a project to standardize variable names, which reflect their usage, but semantics add another layer of expressiveness and control over the exposed resources. 
Let's say you have a material library, where user can also add new materials to the library. Library shader common resource bindings are made available to the user using semantics defined by the library developers, while the naming remains completely free. 

I wanted to know if user defined semantics for shader resources is a feature that is already in the code base (which I completely missed) or is it planned to have such feature?

Again thank you for everything! :pray:

**Comment by csyonghe** [comment, marked as answer]
Hi there! We currently don't have reflection support for HLSL semantics, but we do have reflection support for user defined attributes.

```
[__AttributeUsage(_AttributeTargets.Var)]
struct MyVarAttribute
{
}

[MyVar]
uniform int myVar;
```

You can query for `[MyVar]` with these reflection APIs:
```
TypeReflection::getUserAttributeCount()
UserAttribute* getUserAttributeByIndex()
UserAttribute* getUserAttributeByName()
```

Here are the definition ofr the `_AttributeTargets` enum:
```
enum _AttributeTargets
{
    Struct,
    Var,
    Function
};
```

**Reply by Rodousse** [reply]
Hi!
Thank you very much for the answer. This does cover the kind of use case I was looking for!
Even if it would make more sense to be a "HLSL semantic" kind of decoration in my sense, it is a pretty cool option to have (I'm nit picking sorry).
Again thank you for the solution! 😄
